libs/jwst-codec/src/doc/document.rs

```diff
@@ -70,53 +128,132 @@ impl Doc {
-                if !item.is_skip() {
-                    let local_clock = *state_cache
-                        .entry(client_id)
-                        .or_insert_with(|| self.store.get_state(client_id));
-                    let offset = local_clock as i64 - item.clock() as i64;
-
-                    if offset < 0 {
-                        // applying update depends on another update that not exists on local
-                        rest_items.add_item(item.clone());
-                        rest_items.update_missing_sv(client_id, item.clock() - 1);
-                        // TODO: move all items of the current iterator to rest_items
-                        // rest_items.collect_items();
-                    } else {
+
+        let mut clients = items.keys().copied().collect::<Vec<_>>();
+        // Sort Descending
+        clients.sort();
+
+        let mut state = StateVector::from(&items);
+        let mut rest_structs = RestStructs::new();
+
+        let mut cur_client = clients.pop().unwrap();
+        let mut stack = Vec::new();
+        let mut stack_head = items.get_mut(&cur_client).and_then(|refs| refs.pop());
+
+        while let Some(head) = stack_head {
+            let id = *head.id();
+            let local_clock = state.get(&id.client);
+
+            let offset = (local_clock < id.clock) as i64;
+            // update from the same client is missing
+            if offset < 0 {
+                stack.push(head);
+                rest_structs.update_missing_sv(id.client, id.clock - 1);
+                // continue
+                rest_structs.add_stack(stack, &mut items);
+                stack = Vec::new();
+            } else if let Some(dep) = Self::missing(&head, &state) {
+                stack.push(head);
+                match items.get_mut(&dep) {
+                    Some(refs) => {
+                        stack_head = refs.pop();
+                        continue;
+                    }
+                    None => {
+                        rest_structs.update_missing_sv(dep, state.get(&dep));
+                        rest_structs.add_stack(stack, &mut items);
+                        stack = Vec::new();
+                    }
+                }
+            } else if offset == 0 || (offset as u64) < head.len() {
+                let offset = offset as u64;
+                state.set_max(id.client, id.clock + head.len());
+                self.integrate_struct_info(head, offset);
+            }
+
+            if stack.is_empty() {
+                match items.get_mut(&cur_client) {
+                    Some(v) if !v.is_empty() => {
+                        stack_head = v.pop();
+                    }
+                    _ => {
+                        let next_refs = clients
+                            .pop()
+                            .map(|next_client| (next_client, items.get_mut(&next_client).unwrap()));
+
+                        match next_refs {
```

Comment:
Use `if ... let` pattern here

Reply:
the matches has suffix detection, too verbose in `if let` mode